reference-ril/reference-ril.c

```diff
@@ -1980,39 +2007,38 @@ static void requestSendSMS(void *data, size_t datalen, RIL_Token t)
 
 static int configureInterface(const char* ifname, const char *addr)
 {
-    char ip[16];
+    char ip[INET6_ADDRSTRLEN];
     int prefixLen, ret = -1;
 
-    if (2 != sscanf(addr, "%[.0-9]/%d", ip, &prefixLen))
+    ALOGD("%s: ifname=%s, addr=%s", __FUNCTION__, ifname, addr);
+    if (2 != sscanf(addr, "%[.:0-9a-f]/%d", ip, &prefixLen)) {
+        ALOGE("%s: failed to parse address", __FUNCTION__);
         return ret;
+    }
 
-    if (ifc_init())
+    ret = ifc_act_on_address(RTM_NEWADDR, ifname, ip, prefixLen);
+    if (ret < 0) {
+        ALOGE("%s: ifc_act_on_address returns %d", __FUNCTION__, ret);
         return ret;
-
-    if (!ifc_up(ifname)) {
-        if (ifc_set_addr(ifname, inet_addr(ip)) ||
-            ifc_set_prefixLength(ifname, prefixLen)) {
-            ifc_down(ifname);
-        } else {
-            ret = 0;
-        }
     }
 
-    ifc_close();
+    ret = ifc_enable(ifname);
+    if (ret < 0) {
+        ALOGE("%s: ifc_enable returns %d", __FUNCTION__, ret);
+        ifc_clear_addresses(ifname);
+        return ret;
+    }
 
-    return ret;
+    return 0;
 }
 
-static int deconfigureInterface(const char* ifname)
+static void deconfigureInterface(const char* ifname)
 {
-    int ret;
-
-    if (ifc_init())
-        return -1;
-
-    ret = ifc_down(ifname);
+    ifc_init();
+    ifc_down(ifname);
     ifc_close();
```

Comment:
Sure, let's use ifc_disable(), and since ifc_disable() already clears IPv4 addresses, I'll change the below line to ifc_clear_ipv6_addresses().